### ISSUE TITLE
[Bug?] Non zero variance of output distribution on performing inference on training-data

### ISSUE THREAD

**Issue by ishank-juneja** [cornellius-gp/gpytorch#1821]
# 🐛 Bug

After training a model with train-data `train_x` as a sanity check on the model I am performing inference on the training data itself. I was expecting the means to be precisely my `train_y` labels and the covariance_matrix to be a matrix full of zeros (I am using an RBF kernel so variance at train-points should be 0).

## To reproduce

```python
# Number of train samples
nsamps = 1000
# Fix seed for reproducability
np.random.seed(10)
# Joint input space tensor (u_t, x_t) to hold all inputs and trajectories
train_x = torch.tensor(np.random.uniform(low=-1.0, high=1.0, size=(nsamps, 4))).float()
# A and B matrices
A = torch.tensor([[1., 0.],
                  [0., 1.]])
B = torch.tensor([[-0.2, 0.1],
                 [0.2, -0.1]])
torch.manual_seed(0)

# 1D Output states starting time index 1, no observation noise
train_y = torch.zeros(nsamps, 2)

# Obtain the output states $(x_{t+1, 1}, x_{t+1, 2})$
for i in range(nsamps):
    # Get the output next state
    x_next = torch.matmul(A, train_x[i, 2:4]) + torch.matmul(B, train_x[i, 0:2])
    # No observation noise added
    train_y[i, :] = x_next

dataset = torch.cat([train_x, train_y], dim=1)

class ExactGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super().__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(ard_num_dims=4))

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)


likelihood1 = gpytorch.likelihoods.GaussianLikelihood()
model1 = ExactGPModel(train_x, train_y[:, 0], likelihood1)

likelihood2 = gpytorch.likelihoods.GaussianLikelihood()
model2 = ExactGPModel(train_x, train_y[:, 1], likelihood2)

# Collect the submodels in an IndependentMultiOutputGP, and the respective likelihoods in a MultiOutputLikelihood
model = gpytorch.models.IndependentModelList(model1, model2)
likelihood = gpytorch.likelihoods.LikelihoodList(model1.likelihood, model2.likelihood)
mll = gpytorch.mlls.SumMarginalLogLikelihood(likelihood, model)

# Perform Model Traning
training_iterations = 50

# Find optimal model hyperparameters
model.train()
likelihood.train()

# Use the Adam optimizer
optimizer = torch.optim.Adam([
    {'params': model.parameters()},  # Includes all submodel and all likelihood parameters
], lr=0.1)

for i in range(training_iterations):
    optimizer.zero_grad()
    output = model(*model.train_inputs)
    loss = -mll(output, model.train_targets)
    loss.backward()
    if((i + 1) % 5 == 0):
      print('Iter %d/%d - Loss: %.3f' % (i + 1, training_iterations, loss.item()))
    optimizer.step()

# Performing inference on the training data points themselves
with torch.no_grad(), gpytorch.settings.fast_pred_var():
  # Inference outputs 2 seperate distributions 
  # one for output 1 and the other for output 2
  prediction_dist1 = likelihood(model.models[0](train_x))
  prediction_dist2 = likelihood(model.models[1](train_x))

print(prediction_dist1[0].mean)
print(prediction_dist1[0].covariance_matrix)
print(prediction_dist2[0].mean)
print(prediction_dist2[0].covariance_matrix)
print(train_x)
```

** Stack trace/error message **
```
/usr/local/lib/python3.7/dist-packages/gpytorch/utils/linear_cg.py:278: UserWarning: An output with one or more elements was resized since it had shape [11], which does not match the required output shape [1, 11].This behavior is deprecated, and in a future PyTorch release outputs will not be resized unless they have zero elements. You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0). (Triggered internally at  ../aten/src/ATen/native/Resize.cpp:23.)
  curr_conjugate_vec,
Iter 5/50 - Loss: 0.649
Iter 10/50 - Loss: 0.424
Iter 15/50 - Loss: 0.197
Iter 20/50 - Loss: -0.037
Iter 25/50 - Loss: -0.285
Iter 30/50 - Loss: -0.532
Iter 35/50 - Loss: -0.790
Iter 40/50 - Loss: -1.052
Iter 45/50 - Loss: -1.298
Iter 50/50 - Loss: -1.557

Model 1 Mean and Variance
tensor([-0.0006, -0.0006, -0.0006, ... ,-0.0006],
       requires_grad=True)
tensor([[0.3574, 0.3142, 0.3275,  ..., 0.2934, 0.3189, 0.2843],
        [0.3142, 0.3574, 0.3045,  ..., 0.3240, 0.3275, 0.2961],
        [0.3275, 0.3045, 0.3574,  ..., 0.2697, 0.3163, 0.2545],
        ...,
        [0.2934, 0.3240, 0.2697,  ..., 0.3574, 0.3367, 0.2992],
        [0.3189, 0.3275, 0.3163,  ..., 0.3367, 0.3574, 0.2863],
        [0.2843, 0.2961, 0.2545,  ..., 0.2992, 0.2863, 0.3574]],
       grad_fn=<AddBackward0>)

Model 2 Mean and Variance
tensor([-0.0011, -0.0011, -0.0011, ..., -0.0011],
       requires_grad=True)
tensor([[0.3833, 0.3587, 0.3453,  ..., 0.2573, 0.3124, 0.2836],
        [0.3587, 0.3833, 0.3463,  ..., 0.2657, 0.3188, 0.2918],
        [0.3453, 0.3463, 0.3833,  ..., 0.2021, 0.2769, 0.2286],
        ...,
        [0.2573, 0.2657, 0.2021,  ..., 0.3833, 0.3549, 0.3162],
        [0.3124, 0.3188, 0.2769,  ..., 0.3549, 0.3833, 0.3115],
        [0.2836, 0.2918, 0.2286,  ..., 0.3162, 0.3115, 0.3833]],
       grad_fn=<AddBackward0>)
Original Train Data different from means
tensor([[ 0.0629,  0.7020],
        [-0.6583,  0.5755],
        [ 0.4207,  0.8568],
        ...,
        [-0.6610, -0.8443],
        [-0.1676, -0.4164],
        [-0.2469, -0.3353]])
```

## Expected Behavior

I expected the first column of train_y to be identical to the means output by Model 1 and the second column of train_y to be identical to the means output by the Model 2. Instead the means output by the models in these cases are just constant values (the sonstant values -0.0006 and -0.0011 output here are same as the mean value learned by the model for the constant_mean prior as confirmed by running `model.mean_module.constant`)

Also I expected the covariance matrices associated with the inferred posterior to have all entries close to or equal to 0.0 since for an RBF kernel and for GPs more generally, variance at test points is 0. Instead, the variance entries here are finite values (near about 0.3) that I am unable to interpret.  

## System information

**Please complete the following information:**
- GPyTorch Version (1.5.1)` 
- PyTorch Version (1.10.0+cu111)`
- Google colab notebook: [https://colab.research.google.com/drive/1R-kLiPOwfyTra8MvupQ7tkAoJyjCI77_?usp=sharing](https://colab.research.google.com/drive/1R-kLiPOwfyTra8MvupQ7tkAoJyjCI77_?usp=sharing)

## Additional context
I am working on a linear-dynamics regression toy-example with Gaussian Processes and was getting larger than expected variance values on performing inference on a test trajectory (despite the fact that training seems to go through correctly based on loss values), so I decided to do this sanity check (inference on train data) on the trained GP model.

**Comment by ishank-juneja** [comment]
Issue closed, model and likelihood were not put in eval mode.
After putting in eval mode, the variance values obtain are small.
